File: admin/lib/data/repositories/credits/credit_request_model.dart
/// A shopper's request to the admin for more try-on credits.
class CreditRequestModel {
  final String id;
  final String userId;
  final String userName;
  final String userEmail;
  final int amount;
  final String message;
  final String status; // pending | approved | rejected
  final int grantedAmount;
  final String adminNote;
  final DateTime? createdAt;

  CreditRequestModel({
    required this.id,
    this.userId = '',
    this.userName = '',
    this.userEmail = '',
    this.amount = 0,
    this.message = '',
    this.status = 'pending',
    this.grantedAmount = 0,
    this.adminNote = '',
    this.createdAt,
  });

  bool get isPending => status == 'pending';

  static int _int(dynamic v) => (num.tryParse((v ?? 0).toString()) ?? 0).toInt();

  factory CreditRequestModel.fromJson(Map<String, dynamic> d) =>
      CreditRequestModel(
        id: (d['id'] ?? d['_id'] ?? '').toString(),
        userId: (d['userId'] ?? '').toString(),
        userName: (d['userName'] ?? '').toString(),
        userEmail: (d['userEmail'] ?? '').toString(),
        amount: _int(d['amount']),
        message: (d['message'] ?? '').toString(),
        status: (d['status'] ?? 'pending').toString(),
        grantedAmount: _int(d['grantedAmount']),
        adminNote: (d['adminNote'] ?? '').toString(),
        createdAt: DateTime.tryParse(
            (d['createdAt'] ?? d['CreatedAt'] ?? '').toString()),
      );
}

File: admin/server/src/models/creditRequest.model.js
const mongoose = require('mongoose');
const { baseSchemaOptions } = require('../utils/baseSchema');

/**
 * A shopper's request to the admin for more try-on credits. The admin sees these
 * in the panel's Credits inbox and can approve (which grants the credits) or
 * reject. One pending request per user at a time (enforced in the controller).
 */
const creditRequestSchema = new mongoose.Schema(
  {
    userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, index: true },
    userName: { type: String, default: '' }, // denormalised for the admin list
    userEmail: { type: String, default: '' },
    amount: { type: Number, default: 10 }, // how many the shopper asked for
    message: { type: String, default: '' }, // shopper's note to the admin
    status: {
      type: String,
      enum: ['pending', 'approved', 'rejected'],
      default: 'pending',
      index: true,
    },
    grantedAmount: { type: Number, default: 0 }, // what the admin actually gave
    adminNote: { type: String, default: '' },
    handledBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },
    handledAt: { type: Date, default: null },
  },
  baseSchemaOptions
);

creditRequestSchema.index({ status: 1, createdAt: -1 });

module.exports = mongoose.model('CreditRequest', creditRequestSchema, 'creditrequests');
